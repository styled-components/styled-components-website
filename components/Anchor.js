import styled, { css } from 'styled-components'
import rem from '../utils/rem'

import LinkIcon from 'react-octicons-svg/dist/LinkIcon'
import { Header, SubHeader, TertiaryHeader } from './Layout'
import { mobile } from '../utils/media'

const InvisibleAnchor = styled.div.attrs({
  'aria-hidden': true
})`
  position: relative;
  display: block;
  visibility: hidden;
  height: 0;

  top: ${rem(-70)};

  ${mobile(css`
    top: ${rem(-90)};
  `)}
`

const Anchor = styled.a`
  display: none;
  position: absolute;
  left: 0;
  color: inherit;
`

const AnchorIcon = styled(LinkIcon).attrs({
  width: null,
  height: null
})`
  width: ${rem(20)};
  opacity: 0.7;
  margin-top: ${rem(-5)};

  &:hover {
    opacity: 0.9;
  }
`

const AnchorHeader = styled(Header)`
  position: relative;
  margin-left: ${rem(-30)};
  padding-left: ${rem(30)};

  ${mobile(css`
    margin-left: 0;

    /* stylelint-disable-next-line */
    ${Anchor} {
      display: inline-block;
    }
  `)}

  &:hover ${Anchor} {
    display: inline-block;
  }
`

const AnchorSubHeader = AnchorHeader.withComponent(SubHeader)
const AnchorTertiaryHeader = AnchorHeader.withComponent(TertiaryHeader)

const Link = ({ children, level, id }) => {
  let Child = AnchorHeader

  switch(level) {
    case 3:
      Child = AnchorSubHeader
      break
    case 4:
      Child = AnchorTertiaryHeader
      break
    default:
      break
  }

  return (
    <Child>
      <InvisibleAnchor id={id} />

      <Anchor href={`#${id}`}>
        <AnchorIcon />
      </Anchor>

      {children}
    </Child>
  )
}

export default Link
