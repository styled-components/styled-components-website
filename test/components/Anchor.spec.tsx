import React from 'react';
import renderer from 'react-test-renderer';
import Anchor from '../../components/Anchor';

test('Anchor renders correctly', () => {
  const tree = renderer.create(<Anchor />).toJSON();

  expect(tree).toMatchSnapshot();
});
