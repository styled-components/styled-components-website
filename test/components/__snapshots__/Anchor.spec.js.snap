// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Anchor renders correctly 1`] = `
.c0 {
  font-size: 1.7777777777777777rem;
  font-weight: 600;
  font-family: "Avenir Next",-apple-system,BlinkMacSystemFont,"Segoe UI",Helvetica,Arial,sans-serif,"Apple Color Emoji","Segoe UI Emoji","Segoe UI Symbol";
  margin: 1.5em 0 0.5em;
}

.c2 {
  position: relative;
  display: block;
  visibility: hidden;
  height: 0;
  top: -3.888888888888889rem;
}

.c4 {
  display: none;
  color: inherit;
  margin-left: 0.5555555555555556rem;
}

.c5 {
  width: 1.1111111111111112rem;
  opacity: 0.7;
  margin-top: -0.2777777777777778rem;
}

.c5:hover {
  opacity: 0.9;
}

.c1 {
  position: relative;
}

.c1:hover .c3 {
  display: inline-block;
}

@media (max-width:62.5em) {
  .c2 {
    top: -5rem;
  }
}

@media (max-width:62.5em) {
  .c1 .c3 {
    display: inline-block;
  }
}

<h2
  className="c0 c1"
>
  <div
    aria-hidden={true}
    className="c2"
  />
  <a
    className="c3 c4"
    href="#undefined"
  >
    <svg
      className="c5"
      height={null}
      width={null}
    />
  </a>
</h2>
`;
